Resolve #658: Generate non-trivial json and yaml

worr opened the pull request:
In cases where Cloudformation-specific tags were used (like !GetAtt, for
example), sceptre would fail to generate yaml and json output when `sceptre
generate` was called.

Here we use the loaders provided by Amazon's cfn-lint package to properly read
the cloudformation yaml after substitution, so that we can translate it to yaml
and json correctly without error.

A potential issue with this approach is that I used cfn-lint as a library to provide functionality for resolving all of the custom tags in cloudformation. There aren't any guarantees that I'm aware of for the stability of this API, but it will work at the pinned version. In service of this particular problem, I opened up an issue [here](https://github.com/aws-cloudformation/cfn-python-lint/issues/831) about this, since having a cfn encoder/decoder as a library would be super useful for more than just sceptre.

## PR Checklist

- [x] Wrote a good commit message & description [see guide below].
- [x] Commit message starts with `[Resolve #issue-number]`.
- [ ] Added/Updated unit tests.
- [x] Added/Updated integration tests (if applicable).
- [x] All unit tests (`make test`) are passing.
- [x] Used the same coding conventions as the rest of the project.
- [x] The new code passes flake8 (`make lint`) checks.
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

ngfgrant commented:
I have two lines of thinking on this at the moment.

First, we _could_ utilise `SafeLoader` in pyyaml and implement the required custom yaml constructors to parse `!...` which would generate the template even with things like `!Ref`. This would resolve the generate issue and not introduce a new dependency. We'd also be able to quickly add any missing yaml constructors as appropriate. It is a naive way of doing things but would generate a template.

An example of what I mean is described here: https://stackoverflow.com/questions/41089065/loading-cloudformation-yaml-using-python and there is a project that looks to use this approach: 
https://github.com/awslabs/aws-cfn-template-flip.

The `sceptre generate` command just calls `sceptre.template.body()` Therefore, the functionality for this PR should probably reside in `sceptre/template.py` maybe as a private method something like `_render_template_as_valid_json()` that `template.body()` can utilise i.e.

```python

# https://github.com/cloudreach/sceptre/blob/master/sceptre/template.py#L67

   @property
    def body(self):
        """
        Represents body of the CloudFormation template.
        :returns: The body of the CloudFormation template.
        :rtype: str
        """
        if self._body is None:
            file_extension = os.path.splitext(self.path)[1]
...

        return self._render_template_as_valid_json(self._body)
```

My second line of thinking is that if we do choose to introduce `cfn_lint` as a dependency then I think it would be better to fully utilise the linter i.e. fail to deploy if linter fails; we could trigger the linter on every command. I have a few concerns with this:

- sceptre as a deployment tool starts to become a development tool i.e. scope increase see: #683 
- the task of linting cfn properly is pretty complicated and the package makes no guarantee it is correct. This could end up frustrating our users. i.e what is the point of providing a linter unless it is 100% accurate.
- we will be playing catch up/relying on `cfn-lint` being up to date.
- thinking ahead to Azure and GCP support we would need to offer matching functionality to keep a consistent API/feature parity.

If we did decide to use the linter fully then we could do something like

```python
# sceptre/template.py

@staticmethod
def lint():
     ... implement linting logic

```

On balance, I am not convinced that using the package suggested is necessarily a good idea, even though it might work. Probably a custom SafeLoader would do the trick.

craighurley commented:
The very first line in the readme is:

> Sceptre is a tool to drive CloudFormation

... I think including a linter goes too far for sceptre.

We already use `cfn-lint` separately and use custom ignore filters for particular templates that have some oddities.

worr commented:
I don't particular mean to expand the scope of what sceptre does – that's definitely not within my purview anyway. I just selected that package since it's maintained by Amazon, and provides a loader that I could reuse.

That said, I'm happy to revisit this and implement this with a custom loader if that's what y'all are looking for.

worr commented:
Hey @ngfgrant!

So I finally got time to revisit this. As recommended, I implemented this without pulling in extra deps, just writing a custom loader that can handle and translate the appropriate tags. I also added an integration test as had been pointed out earlier.

I did opt to keep this in `sceptre.cli.helpers`, since this code is only called from `write`, which is in the same file and the entrypoint for writing out these templates. I didn't think it made too much sense to move this into templates, so far from where it's called. If there's a better place to put the constants for the tag names, I'm happy to move them.

ngfgrant commented:
Amazing thanks will review and get this in 2.1.4.

ngfgrant commented on sceptre/cli/helpers.py:
```diff
@@ -268,3 +269,69 @@ def default(self, item):
+]
+
+
+def getatt_constructor(loader, node):
```
We could probably make this a private function?

ngfgrant commented on sceptre/cli/helpers.py:
```diff
@@ -268,3 +269,69 @@ def default(self, item):
         return str(item)
+
+
+FNS = [
```
Could you rename to something like `CFN_FUNCTIONS`

ngfgrant commented on sceptre/cli/helpers.py:
```diff
@@ -268,3 +269,69 @@ def default(self, item):
+    if isinstance(node.value, six.text_type):
+        return node.value.split('.', 1)
+    elif isinstance(node.value, list):
+        ret = loader.construct_sequence(node)
```
What does ret stand for?

ngfgrant commented on sceptre/cli/helpers.py:
```diff
@@ -268,3 +269,69 @@ def default(self, item):
+        raise ValueError("Fn::GetAtt only supports string or list values")
+
+
+def tag_constructor(loader, tag_suffix, node):
```
Could also be a private function (`_tag_constructor`)

ngfgrant commented on sceptre/cli/helpers.py:
```diff
@@ -268,3 +269,69 @@ def default(self, item):
+    'Transform',
+]
+
+TAGS = [
```
Could you rename to something like `CFN_TAGS`

ngfgrant commented on sceptre/cli/helpers.py:
```diff
@@ -268,3 +269,69 @@ def default(self, item):
+
+def tag_constructor(loader, tag_suffix, node):
+    if tag_suffix not in FNS and tag_suffix not in TAGS:
+        raise ValueError("Bad tag: !{tag_suffix}".format(tag_suffix=tag_suffix))
```
We could go a step further and show the list of allowed tags in the error message which might be more helpful?

worr commented:
@ngfgrant I've made allt he requested changes! Let me know if there's anything else you need before merge!
